### PR TITLE
fix(localize): render ICU placeholders in extracted translation files

### LINKED ISSUE #38437: New localize extractor tool: incorrect source/target with plural forms
# 🐞 bug report

### Affected Package

The issue seems to be caused by package @angular/localize

### Is this a regression?

No because it's a new tool, but yes because the old extractor worked.

### Description

When using the new extractor tool introduced in https://github.com/angular/angular/pull/32912 (`./node_modules/.bin/localize-extract`), it seems to create an XLIFF 1.2 file with invalid `source` when using plural in templates. For example with the following template:

```
<span i18n>Updated {minutes, plural, =0 {just now} =1 {one minute ago} other {{{minutes}} minutes ago}}</span>
```

It creates this element

```
<trans-unit id="5a134dee893586d02bffc9611056b9cadf9abfad" datatype="html">
  <source>{VAR_PLURAL, plural, =0 {just now} =1 {one minute ago} other {{INTERPOLATION} minutes ago}}</source>
  <context-group purpose="location">
    <context context-type="sourcefile">../app/app.component.html</context>
    <context context-type="linenumber">1</context>
  </context-group>
</trans-unit>
```

If we use the `source` as a `target`:

```
<trans-unit id="5a134dee893586d02bffc9611056b9cadf9abfad" datatype="html">
  <source>{VAR_PLURAL, plural, =0 {just now} =1 {one minute ago} other {{INTERPOLATION} minutes ago}}</source>
  <target>{VAR_PLURAL, plural, =0 {just now} =1 {one minute ago} other {{INTERPOLATION} minutes ago}}</target>
  <context-group purpose="location">
    <context context-type="sourcefile">../app/app.component.html</context>
    <context context-type="linenumber">1</context>
  </context-group>
</trans-unit>
```

Then we can't build the application:

```
Error: Unexpected character "EOF" (Do you have an unescaped "{" in your template? Use "{{ '{' }}") to escape it.)
Error: Invalid ICU message. Missing '}'.
```

## 🔬 Minimal Reproduction

https://github.com/Chocobozzz/angular-test

## 🔥 Exception or Error
<pre><code>Error: Unexpected character "EOF" (Do you have an unescaped "{" in your template? Use "{{ '{' }}") to escape it.)
Error: Invalid ICU message. Missing '}'.
</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>Angular CLI: 10.1.0-next.4
Node: 12.18.3
OS: linux x64

Angular: 10.1.0-next.5
... animations, common, compiler, compiler-cli, core, forms
... localize, platform-browser, platform-browser-dynamic, router
... service-worker
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.1001.0-next.4
@angular-devkit/build-angular     0.1001.0-next.4
@angular-devkit/build-optimizer   0.1001.0-next.4
@angular-devkit/build-webpack     0.1001.0-next.4
@angular-devkit/core              10.1.0-next.4
@angular-devkit/schematics        10.1.0-next.4
@angular/cli                      10.1.0-next.4
@ngtools/webpack                  10.1.0-next.4
@schematics/angular               10.1.0-next.4
@schematics/update                0.1001.0-next.4
rxjs                              6.6.2
typescript                        3.9.7
webpack                           4.44.1
</code></pre>

**Anything else relevant?**

The issue seems to be `{INTERPOLATION}`

### PR DESCRIPTION
**Pull request by petebacondarwin**
Previously placeholders were only rendered for dynamic interpolation
expressons in `$localize` tagged strings. But there are also potentially
dynamic values in ICU expressions too, so we need to render these as
placeholders when extracting i18n messages into translation files.

Fixes #38437

### CODE REVIEW

File: packages/compiler/test/render3/view/i18n_spec.ts [hunk 1]
```diff
@@ -424,6 +424,15 @@ describe('serializeI18nMessageForLocalize', () => {
...
   });
 
+  it('should serialize ICU with nested interpolation for `$localize()`', () => {
+    expect(serialize('{age, plural, 10 {<b>ten</b>} other {{{age}} years old}}')).toEqual({
+      messageParts: [
+        '{VAR_PLURAL, plural, 10 {{START_BOLD_TEXT}ten{CLOSE_BOLD_TEXT}} other {{INTERPOLATION} years old}}'
+      ],
+      placeHolders: []
+    });
+  });
```
Comment (AndrewKushnir) [inline packages/compiler/test/render3/view/i18n_spec.ts:434]: Should we also add similar test with interpolations in nested ICUs?

File: packages/localize/src/tools/src/extract/translation_files/icu_parsing.ts [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+  const pieces = new IcuPieces();
+  const braces = /[{}]/g;
+
+  let lastPos = 0;
+  let match: RegExpMatchArray|null;
+  while (match = braces.exec(text)) {
+    if (match[0] == '{') {
+      state.enterBlock();
+    } else {
+      state.leaveBlock()!;
```
Comment (AndrewKushnir) [inline packages/localize/src/tools/src/extract/translation_files/icu_parsing.ts:63]:
```suggestion
      state.leaveBlock();
```
May be also leave a quick comment that the `else` block handles the `}` char only, so if we hit that case, we are leaving a block.

**Review by AndrewKushnir — APPROVED** [review]
LGTM, great work @petebacondarwin 👍

File: packages/compiler/test/render3/view/i18n_spec.ts (hunk 1 above)
Comment (petebacondarwin) [reply packages/compiler/test/render3/view/i18n_spec.ts:434]: Sure

**Pull request closed without merging**